tests/framework/prototype.py

```diff
@@ -0,0 +1,245 @@
+        expected_val = search_path(cr, "spec.headGroupSpec.template.spec.containers.0.name".split('.'))
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace = namespace, label_selector='rayNodeType=head')
+        assert(headpods.items[0].spec.containers[0].name == expected_val)
+
+class EasyJobRule(Rule):
+    def assertRule(self, cr=None, namespace='default'):
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace = namespace, label_selector='rayNodeType=head')
+        headpodName = headpods.items[0].metadata.name
+        rtn = os.system("kubectl exec {} -- python -c \"import ray; ray.init(); print(ray.cluster_resources())\""
+                        .format(headpodName))
+        assert(rtn == 0)
+
+class RayClusterAddCREvent(CREvent):
+    def exec(self):
+        client.CustomObjectsApi().create_namespaced_custom_object(
+            group = 'ray.io',version = 'v1alpha1', namespace = self.namespace, plural = 'rayclusters', body = self.cr,
+            pretty = 'true')
+
+    def wait(self):
+        def check_pod_running(pods) -> bool:
+            for pod in pods:
+                if pod.status.phase != 'Running':
+                    return False
+            return True
+        start_time = time.time()
+        expected_head_pods = search_path(self.cr, "spec.headGroupSpec.replicas".split('.'))
+        expected_worker_pods = search_path(self.cr, "spec.workerGroupSpecs.0.replicas".split('.'))
+        # Wait until:
+        #   (1) The number of head pods and worker pods are as expected.
+        #   (2) All head pods and worker pods are "Running".
+        for _ in range(self.timeout):
+            headpods = client.CoreV1Api().list_namespaced_pod(namespace = self.namespace,
+                                                              label_selector='rayNodeType={}'.format('head'))
+            workerpods = client.CoreV1Api().list_namespaced_pod(namespace = self.namespace,
+                                                                label_selector='rayNodeType={}'.format('worker'))
+            if (len(headpods.items) == expected_head_pods and len(workerpods.items) == expected_worker_pods
+                    and check_pod_running(headpods.items) and check_pod_running(workerpods.items)):
+                logger.info("--- %s seconds ---" % (time.time() - start_time))
+                return
+            time.sleep(1)
+        raise Exception("wait() timeout")
+
+# TestSuite
+class GeneralTestCase(unittest.TestCase):
+    def __init__(self, methodName, images, crEvent):
+        super(GeneralTestCase, self).__init__(methodName)
+        self.crEvent = crEvent
+        self.images = images
+
+    def runTest(self):
+        logging.info("Prepare KinD cluster")
+        delete_kind_cluster()
+        create_kind_cluster()
+        install_crd()
+        download_images(self.images)
+        kind_load_images(self.images)
+        install_kuberay_operator()
+        config.load_kube_config()
```

Comment:
With unitTest, I think it's more idiomatic to put lines 216-223 into a `setUp` method.

Reply:
For symmetry, you might also want a teardown method that deletes the kind cluster.

By the way, why are we choosing to delete and recreate kind clusters between test cases?
Does that incur a large overhead? Would it make sense to re-use some of the infra between multiple test cases?

Reply:
1. Added `setUp` and `tearDown` functions in https://github.com/ray-project/kuberay/pull/605/commits/893b38a2482ca53d340bb7e02f640263608ce0e3.

2. "why are we choosing to delete and recreate kind clusters between test cases?"
The reason is that cluster recreation is easy to implement. I have avoided the cluster recreation in https://github.com/ray-project/kuberay/pull/605/commits/893b38a2482ca53d340bb7e02f640263608ce0e3. The execution time for 6 JSON patches reduces from 640s to 419s.